[clang-tidy] Portability Avoid Unprototyped Functions Check

isuckatcs opened the pull request:
This check is inspired by a recent [issue](https://bugs.ruby-lang.org/issues/20908) that happened with Ruby. 

Ruby supports native extensions written in C, and the compilation of those extensions happens on the user side. The source code of the Ruby VM takes advantage of the fact that in C, `void foo()` is a function that can take any number of parameters, however this is no longer true as of C23. As a result users started facing issues when they tried compiling native extensions with C23 flags.

I think the check might be useful for libraries or projects that have an exposed API to their users to give a heads-up that working with unprototyped functions is not fully portable.

The check detects if unprototyped function types are used in the source code.

For example:

```c
  void foo();         // Bad: unprototyped function declaration
  void foo(void);     // Good: a function declaration that takes no parameters

  void (*ptr)();      // Bad: pointer to an unprototyped function
  void (*ptr)(void);  // Good: pointer to a function that takes no parameters
```

Before C23 `void foo()` means a function that takes any number of parameters, so the following snippet is valid.

```c
  // -std=c17
  void foo();

  int main() {
    foo(1, 2, 3);

    return 0;
  }
```

Starting from C23 however, ``void foo()`` means a function that takes no parameters, so the same snippet is invalid.

```c
  // -std=c23
  void foo();

  int main() {
    foo(1, 2, 3);
  //    ^ error: too many arguments to function call, expected 0, have 3

    return 0;
  }
```

Similarly a pointer to an unprototyped function binds to any function before C23, so the following snippet is considered valid.

```c
  // -std=c17
  void foo(int x, int y);

  int main() {
    void (*ptr)() = &foo;

    return 0;
  }
```

From C23 however, the smae pointer will only bind to functions that take no parameters.

```c
  // -std=c23
  void foo(int x, int y);

  int main() {
    void (*ptr)() = &foo;
    //    ^ error: incompatible function pointer types initializing 'void (*)(void)' with an expression of type 'void (*)(int, int)'

    return 0;
  }
```

Some codebases might explicitly take advantage of the pre-C23 behavior, but these codebases should also be aware that
their solution might not be fully portable between compilers.

isuckatcs committed:
[clang-tidy] Portability Avoid Unprototyped Functions Check

isuckatcs commented:
Oh, I missed that thread completely.

My concern about `-Wstrict-prototypes` is that it's in Clang only. I like to think of Clang-Tidy as a standalone linter, which can be used even though the code is being compiled with MSVC or GCC.

In GCC `-Wstrict-prototypes` is not particularly descriptive in it's warnings, as [this example](https://godbolt.org/z/oa3o6f8Eh) showcases it, and in MSVC I'm not aware of any similar warnings.

By having the check in Clang-Tidy, setups that use a different compiler, but also use Clang-Tidy as a linter could still take adavantage of the warning.

vbvictor commented:
> In GCC -Wstrict-prototypes is not particularly descriptive in it's warnings

If you have `-Wstrict-prototypes` enabled in GCC, clang-tidy can show it's corresponding warning from with clang's message if you enable check `clang-diagnostic-strict-prototypes`. https://godbolt.org/z/e9cfsG4EM. But it only works if `-Wstrict-prototypes` is already enabled in `compile_commands.json` or other place.

Now I wonder if we could make a flag for `clang-tidy` to force it to think as if all `clang` diagnostics are enabled, so then we could just write `--checks=clang-diagnostic-strict-prototypes` and have this check enabled regardless of `-Wstrict-prototypes` precense.

carlosgalvezp commented:
You can probably also add the -Wstrict-prototypes to the ExtraArgs field in the .clang-tidy file.

localspook commented:
Hold up — this wasn't obvious to me (I guess because I've never used `ExtraArgs` in the config file), but this is genuinely a pretty convenient solution! Expose a clang warning as a clang-tidy check, all without messing with the build system. I think we should document it!

isuckatcs commented:
> You can probably also add the -Wstrict-prototypes to the ExtraArgs field in the .clang-tidy file.

This still leaves us with the other problem. `-Wstrict-prototypes` doesn't emit anything with `-std=c23`, but for portability reasons we should still require the user to write `void foo(void)` instead of `void foo()` regardless of the standard.

isuckatcs commented:
> So if you have config like this
>
> ```yaml
> # .clang-tidy
> Checks: clang-diagnostic-strict-prototypes
> ExtraArgs: ["-Wstrict-prototypes", "-W...", ...]
> ```

IIUC the `clang-diagnostic-strict-prototypes` check on its own doesn't enable the warning, but what if it did?

I think we could tweak how we handle `clang-diagnostic-<warning>` checks, so `--checks=clang-diagnostic-strict-prototypes` is enough to enable the warning without having to create a `.clang-tidy` file with `ExtraArgs`.

localspook commented:
For code that compiles as C23 but doesn't want to use C23 features to stay compatible with earlier versions, clang provides the [`-Wpre-c23-compat`](https://clang.llvm.org/docs/DiagnosticsReference.html#wpre-c23-compat) warning. However, it seems that warning [doesn't cover your case](https://godbolt.org/z/cEh8n4sPc) (it seems to only warn on language features that are *invalid* in previous versions, not *change meaning* in previous versions). I wonder if we should extend that warning?

isuckatcs committed:
[clang-tidy] Add `clang-additional-diagnostic-*` checks that enable clang warnings

Prior to this change to enable additional clang diagnostics, one had to
specify extra arguments on either the command line, or in the config
file.

This meant invocations like `clang-tidy ... -- -Wall`, or `clang-tidy
--extra-arg=-Wall ...` as well as config files like the one below.

```
\# .clang-tidy
Checks: ...
ExtraArgs: ["-Wall"]
```

This patch merges this functionality with the commonly used checks
interface, so the invocation for enabling the `-Wall` flag will look
like `clang-tidy --checks="...,clang-additional-diagnostic-all" ...`.

isuckatcs commented on clang-tools-extra/clang-tidy/portability/AvoidUnprototypedFunctionsCheck.cpp:
```diff
@@ -0,0 +1,58 @@
+//===--- AvoidUnprototypedFunctionsCheck.cpp - clang-tidy -----------------===//
```
Seems like I used an outdated `add_new_check.py`. 😅

carlosgalvezp commented:
> IIUC the clang-diagnostic-strict-prototypes check on its own doesn't enable the warning, but what if it did?

I think this topic should be discussed in an RFC outside of this thread. Personally I'm not a big fan of adding one more way to do the same thing. I'd rather improve the documentation to clearly describe _the_ way to do it.

isuckatcs commented:
So extending the warning basically means incorporating the functionality of `-Wstrict-prototypes` into `-Wpre-c23-compat`, right? I also wonder what if we simply just preserve `-Wstrict-prototypes` for C23 as well?

@zygoloid what's your take on this, are we allowed to do any of it on the frontend?
